src/hooks/use-dropdown-menu.tsx

```diff
@@ -0,0 +1,152 @@
+// Imports
+import React, { useState, useRef, createRef, useEffect } from 'react';
+import useUpdateOnlyEffect from './use-update-only-effect';
+
+
+// A custom Hook that abstracts away the listeners/controls for dropdown menus
+export default (itemCount: number) => {
+   // Use state
+   const [isOpen, setIsOpen] = useState<boolean>(false);
+   const [returnFocusToButton, setReturnFocusToButton] = useState<boolean>(false);
+   const [currentFocusIndex, setCurrentFocusIndex] = useState<number | null>(null);
+   const firstRun = useRef(true);
+   
+   
+   // Create refs
+   const buttonRef = useRef<HTMLButtonElement>(null);
+   const itemRefs = useRef([...Array(itemCount)].map(() => createRef<HTMLAnchorElement>()));
+   
+   
+   // Create type guard
+   const isKeyboardEvent = (e: React.KeyboardEvent | React.MouseEvent): e is React.KeyboardEvent => (e as React.KeyboardEvent).key !== undefined;
+   
+   
+   // Handle changes in the state of the menu
+   useEffect(() => {
+       // Return if the ref currently points to a null value
+       if (!itemRefs.current[0].current || !buttonRef.current) {
+           return;
+       }
+
+
+       // Return if this is the first fire of the hook, and update ref
+       if (firstRun.current) {
+           firstRun.current = false;
+           return;
+       }
+
+       
+       // If the menu is currently open focus on the first item in the menu, else return focus to the button if state indicates that is desired behavior
+       if (isOpen) {
+           itemRefs.current[0].current.focus();
+           setCurrentFocusIndex(0);
+       } else if (returnFocusToButton) {
+           buttonRef.current.focus();
+       }
+   }, [isOpen, returnFocusToButton]);
+   
+   
+   // Handle changes in the currently focused element
+   useEffect(() => {
+       if (currentFocusIndex !== null) {
+           itemRefs.current[currentFocusIndex].current!.focus(); // TODO: Optional chaining
+       }
+   }, [currentFocusIndex]);
+   
+   
+   // Create handler function for button click and keyboard event
+   const buttonListener = (e: React.KeyboardEvent | React.MouseEvent) => {
+       // Detect if event was a keyboard event or a mouse event
+       if (isKeyboardEvent(e)) {
+           const { key } = e;
+           
+           if (!(key === 'Tab' || key === 'Shift')) {
+               e.preventDefault();
+           }
+           if (key === 'Enter' || key === ' ') {
+               setIsOpen(true);
+               setReturnFocusToButton(false);
+           } else if (key === 'Tab') {
+               setIsOpen(false);
+               setReturnFocusToButton(false);
+           }
+       } else {
+           setIsOpen(!isOpen);
+           setReturnFocusToButton(true);
+       }
+   };
+   
+   
+   // Create handler function for menu controls via keyboard events
```

Comment:
I think this is slightly more clear about what's going on:

```suggestion
   // Create a function that handles menu logic based on keyboard events that occur on menu items
```